wordpress-mobile/WordPress-Android
# Feature/unified comments paging and states

This PR adds simulated paging and loading states to the Unified Comment list.

Currently, you can test the error behavior by enabling airplane mode. 

It's a bit difficult to test errors without network/db layer, so there is one instance where list behavior is temporarily weird - if you load comments, enter airplane mode, perform pull-to-refresh, and rotate a device - the list will be empty. This is expected.

About differences between our current approach to similar functionality and how it works with Paging 3:
- While the data from endpoint/db passes through the VM, it is only there for a purpose of transformation and addition elements, like sub-headers. VM does not know about the data in the list and does not hold it.
- Because of this, loading/error states of the list are managed internally inside PagingAdapter and delivered to us. I pass those states to the VM and transform them into a UiState.
- Adapter is also responsible for requesting data refresh and retry. VM is not involved in this, and only reacts to the result of this action when the loading/error state of the list changes.
- Paging loading indicator and retry button are managed by a separate `LoadingStateAdapter` adapter that you combine with `PagingDataAdapter`. While loading/error state of paging are delivered to us, we don't need to react to them, unless we want to show a SnackBar or something similiar.

[![Image from Gyazo](https://i.gyazo.com/4fb2c6b657c2df4e65bad2ab050bef77.gif)](https://gyazo.com/4fb2c6b657c2df4e65bad2ab050bef77)

[![Image from Gyazo](https://i.gyazo.com/bcf0dd1984bd3fc521ca1b109c88a631.gif)](https://gyazo.com/bcf0dd1984bd3fc521ca1b109c88a631)


To test:
- Navigate to unified comments.
- Confirm that the initial loading indicator is visible.
- Scroll down the list, confirm that the paging loading indicator is visible (we do preloading, btw!).
- Turn on airplane mode.
- Perform pull-to-refresh and confirm that the Snackbar with the error message is visible.
- Scroll the list down to activate paging, confirm that the retry button and error toast is visible.
- Pressing on the Retry button after disabling airplane mode should continue the load.
- While the airplane mode is enabled, reopen Unified Comments.
- Confirm that the full-screen error message is visible after the initial load fails.

## Regression Notes
1. Potential unintended areas of impact
Functionality is still behind a feature flag.

PR submission checklist:
